### PR TITLE
MINIFICPP-2152 Remove own server and client socket implementations

### PR DESCRIPTION
**Pull request by lordgamez**
- Rewrite SiteToSite and CSite2SiteTests to use Asio library instead of custom socket implementation
- Remove custom socket implementation classes and related utils and tests: ClientSocket.h, DescriptorStream.h, EndianCheck.h, ServerSocket.h, Sockets.h, StreamFactory.h, SecureDescriptorStream.h, TLSServerSocket.h, TLSSocket.h
- Remove ProvenanceReportingTest as it was not part of the test suite, it was failing and was unmaintained
- Remove SecureSocketGetTCPTests as it will be replaced in https://issues.apache.org/jira/browse/MINIFICPP-2154

Depends on https://github.com/apache/nifi-minifi-cpp/pull/1592 and https://github.com/apache/nifi-minifi-cpp/pull/1595

https://issues.apache.org/jira/browse/MINIFICPP-2152

-----------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/src/utils/net/DNS.cpp [hunk 1]
```diff
@@ -127,4 +69,9 @@ nonstd::expected<std::string, std::error_code> reverseDnsLookup(const asio::ip::
   return results->host_name();
 }
 
+std::string getMyHostName() {
+  static const std::string HOSTNAME = asio::ip::host_name();
+  return HOSTNAME;
+}
```
Comment (szaszm) [inline libminifi/src/utils/net/DNS.cpp:75]: Is this static caching actually needed? Is querying the hostname too expensive?

File: libminifi/src/utils/net/Socket.cpp [hunk 2]
```diff
@@ -14,19 +14,15 @@
...
-
 #include "utils/net/Socket.h"
-#include "Exception.h"
-#include <cstring>
-#include <system_error>
+
 #ifdef WIN32
-#ifndef WIN32_LEAN_AND_MEAN
-#define WIN32_LEAN_AND_MEAN
-#endif /* WIN32_LEAN_AND_MEAN */
```
Comment (szaszm) [inline libminifi/src/utils/net/Socket.cpp:25]: Why was this removed? This macro reduces the complexity of windows headers if defined before including, and we're still including a windows header below.

File: libminifi/src/utils/net/Socket.cpp [hunk 3]
```diff
@@ -14,19 +14,15 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 #include "utils/net/Socket.h"
-#include "Exception.h"
-#include <cstring>
```
Comment (szaszm) [inline libminifi/src/utils/net/Socket.cpp:20]: `<cstring>` is still used for `memcpy`.

File: libminifi/include/utils/net/Socket.h
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/utils/net/Socket.h:1]: Updated in 1081ccfb9608f8f9dd7933cf48b66cb684276837

File: libminifi/include/utils/net/Socket.h [hunk 4]
```diff
@@ -23,86 +23,16 @@
 #endif /* WIN32_LEAN_AND_MEAN */
 #include <WinSock2.h>
 #else
-#include <sys/types.h>
-#include <sys/socket.h>
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/utils/net/Socket.h:27]: Updated in 1081ccfb9608f8f9dd7933cf48b66cb684276837

File: libminifi/include/utils/net/DNS.h
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/utils/net/DNS.h:1]: Updated in 1081ccfb9608f8f9dd7933cf48b66cb684276837

File: libminifi/src/utils/net/DNS.cpp
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/src/utils/net/DNS.cpp:1]: Updated in 1081ccfb9608f8f9dd7933cf48b66cb684276837

File: libminifi/src/utils/net/DNS.cpp (hunk 1 above)
Comment (lordgamez) [reply libminifi/src/utils/net/DNS.cpp:68]: I'm not sure how expensive is it, but as it should not change at all, I think caching it wouldn't hurt.

File: libminifi/src/utils/net/Socket.cpp (hunk 2 above)
Comment (lordgamez) [reply libminifi/src/utils/net/Socket.cpp]: Updated in 1081ccfb9608f8f9dd7933cf48b66cb684276837

File: libminifi/src/utils/net/Socket.cpp (hunk 3 above)
Comment (lordgamez) [reply libminifi/src/utils/net/Socket.cpp:20]: '<cstring>` is still there it was only moved a bit lower

File: libminifi/src/utils/net/DNS.cpp (hunk 1 above)
Comment (szaszm) [reply libminifi/src/utils/net/DNS.cpp:68]: Hostname can change at runtime. On Linux, you can use the `hostname` command.

File: libminifi/include/core/state/nodes/DeviceInformation.h
Comment (szaszm) [inline libminifi/include/core/state/nodes/DeviceInformation.h:1]: It would be nice to move the logic to the .cpp file, and avoid polluting other translation units with windows and other heavyweight headers.

File: libminifi/include/utils/net/AsioSocketUtils.h [hunk 5]
```diff
@@ -61,6 +62,64 @@ asio::awaitable<std::tuple<std::error_code>> handshake(SslSocket& socket, asio::
...
+      return false;
+    }
+
+    return true;
+  }
+
+  bool connectTcpSocketOverSsl();
+  bool connectTcpSocket();
+
+  asio::io_context io_context_;
```
Comment (szaszm) [inline libminifi/include/utils/net/AsioSocketUtils.h:116]: The old stream factory could maybe be used to inject a common `io_context`, instead of spawning up unique threads for every single connection.

File: libminifi/src/utils/net/DNS.cpp [hunk 6]
```diff
@@ -15,85 +15,20 @@
  * limitations under the License.
  */
 #include "utils/net/DNS.h"
+
+#ifndef WIN32
+#include <cstring>
+#endif /* WIN32 */
+
```
Comment (szaszm) [inline libminifi/src/utils/net/DNS.cpp:22]: What is this used for?

Comment (lordgamez) [reply libminifi/src/utils/net/DNS.cpp]: I don't think it's used anymore, removed in 9ddaa71b8aecb723fc4c8eaaf49873a5939179f4

File: libminifi/include/utils/net/AsioSocketUtils.h (hunk 5 above)
Comment (lordgamez) [reply libminifi/include/utils/net/AsioSocketUtils.h:116]: The asio::io_context does not spawn new threads, it runs synchronously on the thread it was called, in this case through the socket we pass it to with the read/write operations, it should make no difference using separate io contexts for each connection.

File: libminifi/include/core/Processor.h
Comment (fgerlits) [inline libminifi/include/core/Processor.h:1]: to fix the `clang-tidy` warning, I think it would be better to change the `name` parameter to a `string_view` rather than the `const string&` that `clang-tidy` suggests

File: libminifi/include/utils/net/AsioSocketUtils.h [hunk 7]
```diff
@@ -63,6 +64,64 @@ asio::awaitable<std::tuple<std::error_code>> handshake(SslSocket& socket, asio::
...
+  explicit AsioSocketConnection(SocketData socket_data);
+  int initialize() override;
+  size_t read(std::span<std::byte> out_buffer) override {
+    gsl_Expects(stream_);
+    return stream_->read(out_buffer);
+  }
+  size_t write(const uint8_t *value, size_t len) override {
+    gsl_Expects(stream_);
+    return stream_->write(value, len);
+  }
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/utils/net/AsioSocketUtils.h]: Updated in 9c5203b1db3be04c310bc47be0e1199fb0c85102

File: libminifi/include/core/Processor.h
Comment (fgerlits) [reply libminifi/include/core/Processor.h:1]: you're right, `HTTPClient` had the same problem, but `Processor` does, too

Comment (lordgamez) [reply libminifi/include/core/Processor.h:1]: Updated in 9c5203b1db3be04c310bc47be0e1199fb0c85102

**Review by fgerlits — APPROVED** [review]

**Comment by lordgamez** [comment]
> I think we could also remove the now unused close function from FileUtils.h https://github.com/apache/nifi-minifi-cpp/blob/main/libminifi/include/utils/file/FileUtils.h#L422-L428 but otherwise LGTM 👍

Good point, removed the close function in 8526c33ebf3ae5090b165bccf6b2586c0bd7db62

**Review by martinzink — APPROVED** [review]

File: libminifi/include/utils/net/AsioSocketUtils.h [hunk 8]
```diff
@@ -63,6 +64,64 @@ asio::awaitable<std::tuple<std::error_code>> handshake(SslSocket& socket, asio::
...
+    }
+
+    asio::ip::tcp::endpoint local_endpoint(asio::ip::address::from_string(local_network_interface_), 0);
+    asio::error_code err;
+    socket.open(local_endpoint.protocol(), err);
+    if (err) {
+      logger_->log_error("Failed to open socket on network interface '%s' with the following message: '%s'", local_network_interface_, err.message());
+      return false;
+    }
+    socket.set_option(asio::ip::tcp::socket::reuse_address(true));
```
Comment (szaszm) [inline libminifi/include/utils/net/AsioSocketUtils.h:105]: What's the reason for this `reuse_address(true)` option? I think normally it's used to be able to quickly restart a server program that binds to the same address and port as the old process (of the same software) did. We should avoid `bind` and `reuse_address(true)` for client connections: let closed connections linger in the TIME_WAIT state, and let the dynamic binding of ports assign another random port to the client process.

File: libminifi/src/utils/net/AsioSocketUtils.cpp [hunk 9]
```diff
@@ -44,4 +47,66 @@ asio::ssl::context getSslContext(const controllers::SSLContextService& ssl_conte
...
+    result = connectTcpSocket();
+  }
+  return result ? 0 : -1;
+}
+
+bool AsioSocketConnection::connectTcpSocketOverSsl() {
+  auto ssl_context = utils::net::getSslContext(*socket_data_.ssl_context_service);
+  asio::ssl::stream<asio::ip::tcp::socket> socket(io_context_, ssl_context);
+
+  bindToLocalInterface(socket.lowest_layer());
```
Comment (szaszm) [inline libminifi/src/utils/net/AsioSocketUtils.cpp:67]: As mentioned in the previous comment: client sockets (that initiate a connection) should probably not bind to anything.

File: libminifi/include/utils/net/AsioSocketUtils.h (hunk 8 above)
Comment (lordgamez) [reply libminifi/include/utils/net/AsioSocketUtils.h:105]: This is only used if the `local_network_interface_` is set, used by only the `SiteToSitePeer` class. We have the option in the `SiteToSitePeer` constructor to specify a network interface to bind to and I could only port it this way to asio, although this was restricted to Windows systems on the main branch.

Comment (lordgamez) [reply libminifi/include/utils/net/AsioSocketUtils.h:105]: I'm not sure why `reuse_address` was used here, but I think it can be removed.

Comment (szaszm) [reply libminifi/include/utils/net/AsioSocketUtils.h:105]: I found the original code now, and it used `getifaddrs` to find the IP address of the specified interface, when needed. There doesn't seem to be a similar utility in asio.

How is the code here supposed to achieve this? According to the docs, `asio::ip::address::from_string` converts dotted decimal form IP addresses (or hex notation IPv6 address) to binary form, but it doesn't say anything about converting interface names to interface addresses.

Comment (lordgamez) [reply libminifi/include/utils/net/AsioSocketUtils.h]: Updated in a42e21d6c626774f5b9d38b021727a9b37c9eafe

File: libminifi/src/utils/net/AsioSocketUtils.cpp (hunk 9 above)
Comment (lordgamez) [reply libminifi/src/utils/net/AsioSocketUtils.cpp]: Updated in a42e21d6c626774f5b9d38b021727a9b37c9eafe

File: libminifi/include/utils/net/AsioSocketUtils.h [hunk 10]
```diff
@@ -89,28 +94,71 @@ class AsioSocketConnection : public io::BaseStream {
...
+      inet_ntop(AF_INET, &(ipv4_addr->sin_addr), ip_str, INET_ADDRSTRLEN);
+      address = ip_str;
+    } else if (item_found->ifa_addr->sa_family == AF_INET6) {
+      struct sockaddr_in6 *ipv6_addr = (struct sockaddr_in6 *)item_found->ifa_addr;
+      char ip_str[INET6_ADDRSTRLEN];
+      inet_ntop(AF_INET6, &(ipv6_addr->sin6_addr), ip_str, INET6_ADDRSTRLEN);
+      address = ip_str;
+    } else {
+      logger_->log_error("Invalid IP address family found");
+      return;
```
Comment (szaszm) [inline libminifi/include/utils/net/AsioSocketUtils.h:145]: I think we have a `sockaddr_ntop` helper function in Socket.h to replace this logic.

Comment (lordgamez) [reply libminifi/include/utils/net/AsioSocketUtils.h]: Good point, updated in fca226d812d928c168c05ede1a113fb87e77da25

**Review by szaszm — APPROVED** [review]
